include/mfmg/dealii/amge_host.templates.hpp

```diff
@@ -208,25 +208,64 @@ void lanczos_compute_eigenvalues_and_eigenvectors(
 
 template <typename AgglomerateOperator>
 void anasazi_compute_eigenvalues_and_eigenvectors(
-    unsigned int n_eigenvectors, double tolerance,
+    unsigned int n_eigenvectors,
     boost::property_tree::ptree const &eigensolver_params,
     AgglomerateOperator const &agglomerate_operator,
     dealii::Vector<double> const &initial_guess,
+    std::vector<dealii::Vector<double>> const &lobpcg_vectors,
     std::vector<std::complex<double>> &eigenvalues,
     std::vector<dealii::Vector<double>> &eigenvectors)
 {
-  boost::property_tree::ptree anasazi_params;
-  anasazi_params.put("num_eigenpairs", n_eigenvectors);
-  anasazi_params.put("tolerance", std::max(tolerance, 1e-2));
-  anasazi_params.put("max_iterations",
-                     eigensolver_params.get("max_iterations", 1000));
-
   AnasaziSolver<AgglomerateOperator, dealii::Vector<double>> solver(
       agglomerate_operator);
 
   std::vector<double> real_eigenvalues;
+  std::vector<std::shared_ptr<dealii::Vector<double>>> lobpcg_initial_guess;
+  // If the vectors in scratch_data do not exist or if the size of agglomerate
+  // has changed, the initial guess for LOBPCG is the initial provided by the
+  // user.
+  if ((lobpcg_vectors.size() == 0) ||
+      (lobpcg_vectors[0].size() != initial_guess.size()))
+  {
+    lobpcg_initial_guess.resize(1);
+    lobpcg_initial_guess[0] =
+        std::make_shared<dealii::Vector<double>>(initial_guess);
+  }
+  else
+  {
+    lobpcg_initial_guess.resize(n_eigenvectors);
+    for (unsigned int i = 0; i < n_eigenvectors; ++i)
+      lobpcg_initial_guess[i] =
+          std::make_shared<dealii::Vector<double>>(lobpcg_vectors[i]);
+
+    // If the initial vector has zero entries due to the constraints, we need
+    // to modify the LOPBCG initial guess. Conversely if the LOBPCG initial
+    // guess does have zero entries but the initial vector does not, we set
+    // the entries in the LOPBCG initial guess.
+    unsigned int const eigenvector_size = initial_guess.size();
+    for (unsigned int i = 0; i < eigenvector_size; ++i)
+    {
+      if (initial_guess[i] == 0.)
+      {
+        for (unsigned int j = 0; j < n_eigenvectors; ++j)
+        {
+          (*lobpcg_initial_guess[j])[i] = 0.;
+        }
+      }
+      else
+      {
+        for (unsigned int j = 0; j < n_eigenvectors; ++j)
+        {
+          if ((*lobpcg_initial_guess[j])[i] == 0.)
+          {
+            (*lobpcg_initial_guess[j])[i] = initial_guess[i];
+          }
+        }
+      }
+    }
+  }
   std::tie(real_eigenvalues, eigenvectors) =
-      solver.solve(anasazi_params, initial_guess);
+      solver.solve(eigensolver_params, lobpcg_initial_guess);
```

Comment:
So we update lobpcg_init_guess also in this call?

Reply:
What do you mean? Here we just call the solver and use whatever we have set as initial guess.

Reply:
It's https://github.com/ORNL-CEES/mfmg/pull/216/files#diff-5896bf281196f7deec942b95d48b0a90R612